### ISSUE TITLE
[SelectionDAGBuilder] Bad machine code: MBB has duplicate entries in its successor list.

### ISSUE THREAD

**Issue by JonPsson1** [llvm/llvm-project#84801]
SelectionDAGBuilder::visitSwitch()

```
cat tc_MBBlists.ll 
define void @fun(i32 %CC) {
  switch i32 %CC, label %2 [
    i32 0, label %2
    i32 8, label %2
  ]

2:
  ret void
}

llc -mtriple=s390x-linux-gnu -mcpu=z15 -O3  -disable-cgp-branch-opts -verify-machineinstrs ./tc_MBBlists.ll

*** Bad machine code: MBB has duplicate entries in its successor list. ***
- function:    fun
- basic block: %bb.0  (0x9a46ef8)

*** Bad machine code: MBB has duplicate entries in its predecessor list. ***
- function:    fun
- basic block: %bb.1  (0x9a46ff8)
LLVM ERROR: Found 2 machine code errors.
```
@statham-arm ?

**Comment by statham-arm** [comment]
I'm not sure why you've tagged me – do I know something about this?

I certainly don't know anything about `s390x`. But I see that the same crash happens with this Arm command line:
```
build/bin/llc -print-after-all -mtriple=arm-linux-gnu -O3  -disable-cgp-branch-opts -verify-machineinstrs ./tc_MBBlists.ll
```

It goes away without `-disable-cgp-branch-opts`. But I don't know what that option does.

**Comment by JonPsson1** [comment]
> I'm not sure why you've tagged me – do I know something about this?
> 
> I certainly don't know anything about `s390x`. But I see that the same crash happens with this Arm command line:
> 
> ```
> build/bin/llc -print-after-all -mtriple=arm-linux-gnu -O3  -disable-cgp-branch-opts -verify-machineinstrs ./tc_MBBlists.ll
> ```
> 
> It goes away without `-disable-cgp-branch-opts`. But I don't know what that option does.

I think I found you in the log for SelectionDAGBuilder, so was just checking if you knew anything about it... :) Thanks for taking a look.

This only triggers without the branch optimizations - I think there should be a check somewhere to not add the same successor/predecessor twice to the MBB lists...

**Comment by statham-arm** [comment]
Ah, commit 60b98363c7ed0a549be4d51ee07c32dc2bf47d2f? I'm afraid I found that by following breadcrumbs from the Arm backend, and I don't know any more about SelectionDAGBuilder than I had to learn to fix that one issue.

**Comment by JonPsson1** [comment]
> Ah, commit 60b9836?...
Yeah... Adding some other folks here in case they know more: @yetingk @paulwalker-arm @jasilvanus
